thepracticaldev/dev.to
# Add ahoy + welcome notification tracking

## What type of PR is this? (check all applicable)


## Description

We deployed welcome notifications last cycle, but haven't been able to measure or quantify how effective they are, and which ones are performing better than others.

This PR adds tracking around welcome notifications by making use of the `ahoy` [gem](https://github.com/ankane/ahoy) and [javascript library](https://github.com/ankane/ahoy.js).

This PR does a few things:
- [x] Adds two dependencies for `ahoy` (rails + JS), and their corresponding models, migrations, initializers, and database tables.
- [x] Imports the JS dependency into `base.js.erb` so that it can be used anywhere (although currently, we're only using it in the welcome notifications view).
- [x] Unique `id`s for each `broadcast` notification that is rendered on `/notifications`
- [x] A `trackNotification()` function evaluated in a `<script>` tag, which will track clicks for welcome notifications.

## Related Tickets & Documents
Closes https://github.com/thepracticaldev/dev.to/issues/8099.

## QA Instructions, Screenshots, Recordings

1. First, add or update a `Welcome` broadcast to include a tracking function:
<img width="1450" alt="Screen_Shot_2020-06-17_at_5_51_26_PM" src="https://user-images.githubusercontent.com/6921610/84965558-158f1a80-b0c4-11ea-9988-fc7a7880fcdf.png">

Your broadcast should look something like this:
<img width="1024" alt="Screen Shot 2020-06-17 at 4 09 19 PM" src="https://user-images.githubusercontent.com/6921610/84965569-19bb3800-b0c4-11ea-8775-5b144edd5eb2.png">

2. In your terminal, open up `rails console`. You'll want to find your user, and delete any notifications you already have for the broadcast that you just modified:
```bash
$ rails console
> user = User.where(email: 'put-your-email@here.com')
> user.notifications
> # Delete the preexisting notification here, for example something like this:
> user.notifications.last.destroy
> # Check that you don't have any events already tracked:
> Ahoy::Event.count
> 0
```
3. Manually send yourself a notification from the console:
```bash
# inside of the rails console
> Broadcasts::WelcomeNotification::Generator.call(user.id)
> nil
```
You should see something like this happen in your server logs:
```bash
# bin/startup
> 18:10:41 sidekiq.1   | 2020-06-18T01:10:41.189Z pid=10759 tid=21af class=Notifications::WelcomeNotificationWorker jid=366f055672bee68e2a677af4 INFO: start
> 18:10:41 sidekiq.1   | 2020-06-18T01:10:41.638Z pid=10759 tid=21af class=Notifications::WelcomeNotificationWorker jid=366f055672bee68e2a677af4 elapsed=0.451 INFO: done
```
4. Go to your `/notifications` tab. You should see the click handler appear, as well as an unique `id` on the broadcast element. Click on the link that you jut added tracking to!
<img width="1344" alt="Screen_Shot_2020-06-17_at_6_15_27_PM" src="https://user-images.githubusercontent.com/6921610/84966646-db734800-b0c6-11ea-88ab-a94b3c5dbe6a.png">

5. The click should take you to the page that you expect, and when you go back into your `rails console`, you should see a new `Ahoy::Event` that records the action you just took:

```bash
# inside of the rails console
> Ahoy::Event.count
> 1
> Ahoy::Event.last
> #<Ahoy::Event:0x00007fdadde4bc30
 id: 15,
 name: "Clicked Welcome Notification",
 properties: {"title"=>"welcome_notification_welcome_thread"},
 time: Thu, 18 Jun 2020 01:20:10 UTC +00:00,
 user_id: 11,
 visit_id: 6>
```

_Hooray, you just tracked the click on a welcome notification! 🎉_

## Added tests?


## Added to documentation?

- [x] ~readme~ **added inline docs**

## Are there any post deployment tasks we need to perform?
Once we have shipped this PR, we'll need to add the `trackNotification(e)` to the broadcasts in the welcome notification flow that we want to track; without that, we won't actually be tracking them :)
